### PR TITLE
PUBDEV-5669: Hive import performance improvement

### PR DESCRIPTION
**Pull request by seb-h2o**
https://0xdata.atlassian.net/browse/PUBDEV-5669

Idea is to first use one frame for optimal data retrieval (minimum amount of chunks but with maximum connections for max parallelism) and then rebalance automatically to another frame for optimal data manipulation later.

Also added possibility to use a subselect instead of a temp table as table creation can become very expensive on large tables: the drawback is that we lose some isolation when querying the table, but it should not be a major concern on hive tables (the amount of rows being retrieved is computed first): for now could be activated just for hive.

### CODE REVIEW

**Comment by michalkurka** [comment]
@seb-h2o can you please separate refactoring and functional changes?

This is very hard to review

File: h2o-core/src/main/java/water/fvec/Vec.java [hunk 1]
```diff
@@ -371,12 +371,20 @@ public static Vec makeCon(double x, long len, int log_rows_per_chunk) {
...
   /**
    * Make a new constant vector with minimal number of chunks. Used for importing SQL tables.
    *  @return New constant vector with the given row count. */
   public static Vec makeCon(long totSize, long len) {
-    int safetyInflationFactor = 8;
-    int nchunks = (int) Math.max(safetyInflationFactor * totSize / Value.MAX , 1);
+      return makeCon(totSize, len, VALUE_MAX);
+  }
+
+  public static Vec makeCon(long totSize, long len, int valueMax) {
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/fvec/Vec.java]: What is this method for?

**Comment by seb-h2o** [comment]
@michalkurka: for the split, do you mean 2 commits or 2 PRs?

File: h2o-core/src/main/java/water/fvec/Vec.java (hunk 1 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/fvec/Vec.java]: not used anymore, was used during benchmarks when changing the default magic valueMax.
will remove

File: h2o-core/src/main/java/water/fvec/Vec.java [hunk 2]
```diff
@@ -371,12 +371,20 @@ public static Vec makeCon(double x, long len, int log_rows_per_chunk) {
     return makeCon(x,len,log_rows_per_chunk,true);
   }
 
+  private static int VALUE_MAX = 1<<28; //original = 1<<28 = Value.MAX / 8
   /**
    * Make a new constant vector with minimal number of chunks. Used for importing SQL tables.
    *  @return New constant vector with the given row count. */
   public static Vec makeCon(long totSize, long len) {
-    int safetyInflationFactor = 8;
-    int nchunks = (int) Math.max(safetyInflationFactor * totSize / Value.MAX , 1);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/fvec/Vec.java]: Why do you need to change this? I don't think this is a step in the right direction because you are losing the relationship with Value.MAX. If Value.MAX is changed in the future it won't be reflected here resulting in a bug.

File: h2o-core/src/main/java/water/fvec/Vec.java [hunk 3]
```diff
@@ -394,7 +402,7 @@ public static Vec makeCon(double x, long len, int log_rows_per_chunk, boolean re
 
   public static Vec makeCon(double x, long len, int log_rows_per_chunk, boolean redistribute, byte type) {
     int chunks0 = (int)Math.max(1,len>>log_rows_per_chunk); // redistribute = false
-    int chunks1 = (int)Math.min( 4 * H2O.NUMCPUS * H2O.CLOUD.size(), len); // redistribute = true
+    int chunks1 = (int)Math.min( 4 * H2O.ARGS.nthreads * H2O.CLOUD.size(), len); // redistribute = true
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/fvec/Vec.java]: Good catch!

File: h2o-core/src/main/java/water/fvec/Vec.java (hunk 2 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/fvec/Vec.java]: I'll restore the default to Value.MAX, I was much more intrigued by the so-called safety factor that mainly looks like a magic number without proper rationale outside artificially changing the Value.MAX

File: h2o-core/src/main/java/water/fvec/Vec.java (hunk 3 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/fvec/Vec.java]: there are a bunch like this in the code base: I'm actually thinking about creating a ticket to review usages of NUMCPUS, Runtime.getRuntime().availableProcessors() and so on

**Comment by michalkurka** [comment]
@seb-h2o I would do 2 PRs - one with refactoring one and the second one for changing the behavior.

The point is to see the changes of the original code.

**Comment by seb-h2o** [comment]
cf. https://github.com/h2oai/h2o-3/pull/2667 for refactorings, I'll remove them for this PR soon

**Comment by seb-h2o** [comment]
@michalkurka only feature changes should be visible on this PR now

**Comment by seb-h2o** [comment]
@michalkurka what's the status of that one, are we keeping it for later use?

File: h2o-core/src/main/java/water/fvec/Vec.java (hunk 2 above)
Comment (michalkurka) [reply h2o-core/src/main/java/water/fvec/Vec.java]: I think it is a little confusing to have 2 makeCon() methods that only differ in type long vs int.

This will break existing code that is calling the old method

     int x = 17;
     Vec.makeCon(42.0, x);

Will start behaving different!

Comment (seb-h2o) [reply h2o-core/src/main/java/water/fvec/Vec.java]: fair enough as the int and long versions have a different semantic besides: will rename the new one to sth like makeConN to hint that the size N of the vector is fixed and defined by param nchunks

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 4]
```diff
@@ -171,39 +185,62 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
...
             (long)((float)(catcols+intcols)*numRow*4 //4 bytes for categoricals and integers
                     +(float)bincols          *numRow*1*binary_ones_fraction //sparse uses a fraction of one byte (or even less)
                     +(float)(realcols+timecols+stringcols) *numRow*8); //8 bytes for real and time (long) values
-    final Vec _v;
+
+    final Vec _retrieval_v;
+    final int chunk_size = FileVec.calcOptimalChunkSize(totSize, numCol, numCol * 4,
+            H2O.ARGS.nthreads, H2O.getCloudSize(), false, false);
+    final double rows_per_chunk = chunk_size; //why not numRow * chunk_size / totSize; it's supposed to be rows per chunk, not the byte size
+    final Vec _v = makeCon(0, numRow, (int) Math.ceil(Math.log1p(rows_per_chunk)), false);
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: I would remove the leading underscore from `_v` and `_retrieval_v`. It goes against the coding conventions and this is a good opportunity to fix it.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 5]
```diff
@@ -171,39 +185,62 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
...
+      _retrieval_v = _v;
     }
-    Log.info("Number of chunks: " + _v.nChunks());
+    //if autoRebalance set to true, we first use optimal #chunks for retrieval and then immediately rebalance to optimal #chunks for later processing
+    final boolean autoRebalance = _v != _retrieval_v && Boolean.parseBoolean(System.getProperty(AUTO_REBALANCE_ENABLED, "false"));
+    Log.info("Number of chunks for data retrieval: " + _retrieval_v.nChunks());
+    Log.info("Number of final chunks: " + (autoRebalance ? _v.nChunks() : _retrieval_v.nChunks()));
     //create frame
-    final Key destination_key = Key.make(table + "_sql_to_hex");
+    final Key destination_key = Key.make((table + "_sql_to_hex").replaceAll("\\W", "_"));
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: thank you!

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 6]
```diff
@@ -171,39 +185,62 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
...
+        final Frame retrieval_fr = new SqlTableToH2OFrame(finalTable, needFetchClause, columns, numCol, j, provider)
+                .doAll(columnH2OTypes, _retrieval_v)
+                .outputFrame(tmpKey, columnNames, null);
+
+        if (autoRebalance) {
+          final RebalanceDataSet rds = new RebalanceDataSet(retrieval_fr, destination_key, _v.nChunks());
+          H2O.submitTask(rds);
+          fr = rds.getResult();
+        } else {
+          fr = new Frame(destination_key, retrieval_fr.names(), retrieval_fr.vecs());
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: the logic on lines 230-232 is not straightforward; the code can be simplified:

1) use `destination_key` directly for the retrieval_frame if rebalancing is disabled
2) delete `retrieval_fr` explicitly when rebalance is done

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 7]
```diff
@@ -171,39 +185,62 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
...
+        }
+        Frame.deleteTempFrameAndItsNonSharedVecs(retrieval_fr, fr);
+        _retrieval_v.remove();
         _v.remove();
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: Is Vec `_v` used for anything? I am probably missing something but it looks like it is merely holding number of chunks and it is not used as a Vec.

**Comment by michalkurka** [comment]
@Pscheidl please review especially changes in your code

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 6 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: Yep, good idea, thanks!
I was actually wondering if RebalanceDataSet could rebalance a frame to its own key, but apparently this should work and save the additional frame.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 8]
```diff
@@ -269,26 +306,26 @@ public ConnectionPoolProvider() {} // Externalizable classes need no-args constr
...
-          conPerNode = calculateLocalConnectionCount(MAX_CONNECTIONS, cloudSize, nThreads, nChunks);
+        Integer userMaxConnections = Integer.valueOf(userDefinedMaxConnections);
+        if (userMaxConnections > 0 && userMaxConnections < MAX_CONNECTIONS) {
+          maxConnections = userMaxConnections;
         }
       } catch (NumberFormatException e) {
         Log.info("Unable to parse maximal number of connections: " + userDefinedMaxConnections
-            + ". Falling back to default settings.");
-        conPerNode = calculateLocalConnectionCount(MAX_CONNECTIONS, cloudSize, nThreads, nChunks);
+                + ". Falling back to default settings.");
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: This part was written be me and I'd like to enhance it a little bit. Would it be possible to add the number of connections we're backing up to ? Just concat the `MAX_CONNECTIONS` constraint ?

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 9]
```diff
@@ -304,6 +341,10 @@ private static final int calculateLocalConnectionCount(final int maxTotalConnect
       //Make sure at least some connections are available to a node
       return Math.max(conPerNode, MIN_CONNECTIONS_PER_NODE);
     }
+
+    static int getOptimalConnectionPerNode(final int cloudSize, final short nThreads) {
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: Javadoc + maybe we can keep private modifier in place.  Or is this because of tests ?

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 10]
```diff
@@ -269,26 +306,26 @@ public ConnectionPoolProvider() {} // Externalizable classes need no-args constr
...
       }
+      return maxConnections;
+    }
 
-      return conPerNode;
+    /**
+     * @return Number of connections to an SQL database to be opened on a single node.
+     */
+    static int getMaxConnectionsPerNode(final int cloudSize, final short nThreads, final int nChunks) {
+     return calculateLocalConnectionCount(getMaxConnectionsTotal(), cloudSize, nThreads, nChunks);
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: This seems to be just a delegate method to my old method. Maybe we can call `getMaxConnectionsTotal()! directly in `calculateLocalConnectionCount! - because the task here is to obtain local connections count per node anyway. The call chains get a little bit confusing. Let's keep just one method - in a way it makes sense to call `getMaxConnectionsTotal`in calculateLocalConnectionCount, because in order to calculate the number of local connections, the maximal number of connections must be known.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 11]
```diff
@@ -304,6 +341,10 @@ private static final int calculateLocalConnectionCount(final int maxTotalConnect
       //Make sure at least some connections are available to a node
       return Math.max(conPerNode, MIN_CONNECTIONS_PER_NODE);
     }
+
+    static int getOptimalConnectionPerNode(final int cloudSize, final short nThreads) {
+      return Math.min(nThreads, Math.max(getMaxConnectionsTotal() / cloudSize, MIN_CONNECTIONS_PER_NODE));
```
Comment (Pscheidl) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: I don't see the optimization here. The `Math.max(getMaxConnectionsTotal() / cloudSize, MIN_CONNECTIONS_PER_NODE)`. In many cases, this is similar to what I already have. In some edge cases, it may LOWER the number of connections (falling back to one connection per threads if the calculations result in higher number of connections than there are threads).

I'd image there is a reason behind having one connection per thread, yet some connections may just be in a state of waiting for resources and one thread can actually handle more than one connection while waiting for the others to be in the "ready" state. Is HIVE the reason for this optimization ?

**Comment by Pscheidl** [comment]
Added few questions of my own. What I dislike is the way we calculate connections right now. It has just become too complex in my opinion. The "optimize" regime was there for some reason, added by someone. We changed it now. 

I'd strongly urge we just drop the "optimize" regime in this PR have one working and tested model of resource allocation. When it comes to chunk distribution, our goal is always to have optimal amount spread across all the nodes to achieve maximal amount of parallelism, is that right ? 

I know my commentary does complicate things a little bit, but it could be more benefitial in future. Also, the "optimized" regime is a little bit hidden and too complex for H2O users maybe. Everyone wants things optimized out of the box, right ? Really sounds weird and I don't like how complex it is right now.

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 8 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: sure

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 12]
```diff
@@ -269,26 +306,26 @@ public ConnectionPoolProvider() {} // Externalizable classes need no-args constr
 
     }
 
-    /**
-     * @return Number of connections to an SQL database to be opened on a single node.
-     */
-    static int getMaxConnectionsPerNode(final int cloudSize, final short nThreads, final int nChunks) {
-      int conPerNode;
+    static int getMaxConnectionsTotal() {
```
[the remark this answers is not in the document]
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: done

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 10 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: I've just split the original getMaxConnectionsPerNode, because it was doing 2 different things:
1. reading the *global* property: max_connections param/defaults: this is extracted to getMaxConnectionsTotal().
2. computing the *local* connections needed for a single node to fetch data.
Therefore, I think separation of concerns applies here.
Besides, #1 is also needed by getOptimalConnectionPerNode below.

Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: Oh I see, if you want me to add your existing calculateLocalConnectionCount logic, into this method, then I'll rather do it in the refactoring branch: people don't like when refactorings are mixed with logic changes :)

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 9 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: done

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 11 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: the point of this method is not to calculate the real number of connections used later: it is to make an estimation of the optimal amount of connections needed to retrieve the data, regardless of the size of the data...
The fact is that this "optimized" logic always use more connections than the previous one: the previous logic was creating 1 single chunk (and for this using 1 connection) for every 256MB of data. The current logic is trying to increase parallelism, but also avoiding waiting connections that would have a bad impact on fetching performance.
Basically most of the time consumed when fetching data os in iterating through the resultset, so waiting connections are useless, you just want to have as many parallel active connections as you can sucking up the data, without also having to make too many queries.

Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: let's discuss if more clarification needed

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 4 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: done

Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: removing it for other attributes in this file as well

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 11 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: closed after discussion

**Comment by seb-h2o** [comment]
@Pscheidl , after our discussion we all agree that the optimize flag is a burden, but this is out of scope for this perf task (requires change everywhere down to clients and API): I created a task for this https://0xdata.atlassian.net/browse/PUBDEV-5788

**Comment by seb-h2o** [comment]
@michalkurka , @Pscheidl: I addressed your comments, please have a look

File: h2o-core/src/main/java/water/fvec/Vec.java (hunk 2 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/fvec/Vec.java]: closed

File: h2o-core/src/main/java/water/jdbc/SQLManager.java (hunk 10 above)
Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: closed

**Comment by seb-h2o** [comment]
@michalkurka, @Pscheidl : I believe we've been going through enough review loops on this ticket. All concerns should have been addressed by now, without any update on your side, merging this at the end of the day, thanks.

**Comment by michalkurka** [comment]
@seb-h2o this should NOT be merged yet

File: h2o-core/src/main/java/water/jdbc/SQLManager.java [hunk 13]
```diff
@@ -171,49 +183,71 @@ SELECT ROW_NUMBER() OVER () AS RowNum_, <table>.* FROM <table>
...
     };
-    j.start(work, _v.nChunks());
+    j.start(work, vec.nChunks());
     
     return j;
   }
 
   static class ConnectionPoolProvider extends Iced<ConnectionPoolProvider> {
 
-    private String _url;
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/jdbc/SQLManager.java]: I would rather respect h2o coding conventions

Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: where can I find those? the codebase itself is very inconsistent.
you mentioned in a previous comment: "I would remove the leading underscore from _v and _retrieval_v. It goes against the coding conventions and this is a good opportunity to fix it."
so, I removed leading underscores everywhere in this file.
Do you mean that it should stay for private fields?
Btw, those renaming are not really related with my changes, I'd rather use the refactoring branch (PR 2667) for that.

Comment (michalkurka) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: The conventions are not formalized.

The one rule we follow is to use underscores in the name of class fields. That distinguishes them from method variables. In your case `_v` was a variable, here we are dealing with class fields which already begin with underscore as expected.

Comment (seb-h2o) [reply h2o-core/src/main/java/water/jdbc/SQLManager.java]: changed

**Comment by michalkurka** [comment]
Approved from my point of view, let's wait for @Pscheidl's review before merging. Thx!

**Comment by Pscheidl** [comment]
This code undergoes many changes and sebastien already reacted to my concerns. Sorry about the delay.

Let's do the refactorings we both agreed on (me and @seb-h2o ) in future, as there is no business value to that.

**Comment by Pscheidl** [comment]
I only wonder if we need more tests in SQLManagerTest class. My tests only test min/max connections related to number of CPUs available. 

I think there could be a test telling us the connection pool is as big as we want it to be after this change ? Just a question to ask ourselves before we merge. The test would describe the mechanism behind this new "strategy" very well. And a benchmark would only confirm it.

My recent changes remained untouched and are still valid, this is ensured by the tests.

**Comment by seb-h2o** [comment]
@michalkurka: I've been running more tests on Fri. as I want to get numbers for large tables (100M rows) that I was not able to obtain previously, which led me to find an issue with the latest rebalance logic when using the same key as the frame it's rebalancing.
I finish gathering metrics for 100M rows import+training, comparing rel-wright with this PR with autorebalance enabled or not, will publish all results here and in JIRA today and then ask for merge confirmation.

**Comment by seb-h2o** [comment]
@michalkurka please see "definitive" results up to 100M rows tables in the following link:
https://s3.eu-central-1.amazonaws.com/sebp/sql_benchmark/results/sql_import_results.md

**Comment by michalkurka** [comment]
@seb-h2o great performance report, thank you!

I would consider removing the auto-rebalance option from the code (eventually). Rebalancing categoricals (represented) as strings will be costly - it is most likely better to let the users convert the values to enums and then use the auto-rebalance on the model level (for the most common use case).

I think we can go ahead and merge the PR, wdyt?

**Comment by seb-h2o** [comment]
@michalkurka I agree, after those new tests, I started to consider removing autorebalance as well.
we can merge now, and I can quickly remove the autorebalance in a fast-path PR tmrw.
